src/anomalib/callbacks/metrics.py

```diff
@@ -182,16 +181,37 @@ def _update_metrics(
         image_metric.update(output["pred_scores"], output["label"].int())
         if "mask" in output and "anomaly_maps" in output:
             pixel_metric.to(self.device)
-            pixel_metric.update(torch.squeeze(output["anomaly_maps"]), torch.squeeze(output["mask"].int()))
+            if "masks" in output:
+                self._update_pixel_metrics(pixel_metric, output)
+            else:
+                pixel_metric.update(torch.squeeze(output["anomaly_maps"]), torch.squeeze(output["mask"].int()))
 
     def _outputs_to_device(self, output: STEP_OUTPUT) -> STEP_OUTPUT | dict[str, Any]:
         if isinstance(output, dict):
             for key, value in output.items():
                 output[key] = self._outputs_to_device(value)
         elif isinstance(output, torch.Tensor):
             output = output.to(self.device)
+        elif isinstance(output, list):
+            for i, value in enumerate(output):
+                output[i] = self._outputs_to_device(value)
         return output
 
+    def _update_pixel_metrics(self, pixel_metric: AnomalibMetricCollection, output: STEP_OUTPUT) -> None:
+        """Handle metric updates when the SPRO metric is used alongside other pixel-level metrics."""
+        update = False
+        for name, metric in pixel_metric.items(copy_state=False):
+            if isinstance(metric, SPRO):
+                metric.update(torch.squeeze(output["anomaly_maps"]), output["masks"])
```

Comment:
@djdameln this approach was actually to address this comment https://github.com/openvinotoolkit/anomalib/pull/1686#discussion_r1478718458 . I provided a more detailed explanation in this comment https://github.com/openvinotoolkit/anomalib/pull/1686#discussion_r1479756383

My initial approach was as follows:

```
pixel_metric.update( 
         torch.squeeze(output["anomaly_maps"]), 
         torch.squeeze(output["mask"].int()), 
         masks=torch.squeeze(output["masks"]) if "masks" in output else None, 
     ) 
```

My initial approach can handle all metrics (including SPRO and others) without metric-specific logic in the callbacks.metrics.py file. However, the problem is that we need an additional argument, `_`:

```
def update(self, predictions: torch.Tensor, _: torch.Tensor, masks: torch.Tensor) -> None:
```
So, I changed the solution to avoid using `_` by incorporating metric-specific logic. Is there a better idea to avoid metric-specific logic while also avoiding the use of `_`?